File: docs/core/examples/ruby/accounts.rb
require 'chain'

chain = Chain::Client.new
signer = Chain::HSMSigner.new

asset_key = chain.mock_hsm.keys.create
signer.add_key(asset_key, chain.mock_hsm.signer_conn)

alice_key = chain.mock_hsm.keys.create
signer.add_key(alice_key, chain.mock_hsm.signer_conn)

bob_key = chain.mock_hsm.keys.create
signer.add_key(bob_key, chain.mock_hsm.signer_conn)

chain.assets.create(
  alias: 'gold',
  root_xpubs: [asset_key.xpub],
  quorum: 1,
)

chain.assets.create(
  alias: 'silver',
  root_xpubs: [asset_key.xpub],
  quorum: 1,
)

# snippet create-account-alice
chain.accounts.create(
  alias: 'alice',
  root_xpubs: [alice_key.xpub],
  quorum: 1,
  tags: {
    type: 'checking',
    first_name: 'Alice',
    last_name: 'Jones',
    user_id: '12345',
  }
)
# endsnippet

# snippet create-account-bob
chain.accounts.create(
  alias: 'bob',
  root_xpubs: [bob_key.xpub],
  quorum: 1,
  tags: {
    type: 'savings',
    first_name: 'Bob',
    last_name: 'Smith',
    user_id: '67890',
  }
)
# endsnippet

# snippet list-accounts-by-tag
chain.accounts.query(
  filter: 'tags.type=$1',
  filter_params: ['savings'],
).each do |a|
  puts "Account ID #{a.id} alias #{a.alias}"
end
# endsnippet

# snippet update-account-tags
chain.accounts.update_tags(
  alias: 'alice',
  tags: {
    type: 'money market',
    first_name: 'Alice',
    last_name: 'Jones',
    user_id: '12345',
  },
)
# endsnippet

fund_alice_tx = chain.transactions.build do |b|
  b.issue asset_alias: 'gold', amount: 100
  b.control_with_account account_alias: 'alice', asset_alias: 'gold', amount: 100
end

chain.transactions.submit(signer.sign(fund_alice_tx))

fund_bob_tx = chain.transactions.build do |b|
  b.issue asset_alias: 'silver', amount: 100
  b.control_with_account account_alias: 'bob', asset_alias: 'silver', amount: 100
end

chain.transactions.submit(signer.sign(fund_bob_tx))

# snippet build-transfer
spending_tx = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'alice', asset_alias: 'gold', amount: 10
  b.control_with_account account_alias: 'bob', asset_alias: 'gold', amount: 10
end
# endsnippet

# snippet sign-transfer
signed_spending_tx = signer.sign(spending_tx)
# endsnippet

# snippet submit-transfer
chain.transactions.submit(signed_spending_tx)
# endsnippet

# snippet create-receiver
bob_receiver = chain.accounts.create_receiver(
  account_alias: 'bob'
)
bob_receiver_serialized = bob_receiver.to_json
# endsnippet

# snippet transfer-to-receiver
spending_tx2 = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'alice', asset_alias: 'gold', amount: 10
  b.control_with_receiver(
    receiver: JSON.parse(bob_receiver_serialized),
    asset_alias: 'gold',
    amount: 10
  )
end

chain.transactions.submit(signer.sign(spending_tx2))
# endsnippet

# snippet list-account-txs
transactions = chain.transactions.query(
  filter: 'inputs(account_alias=$1) AND outputs(account_alias=$1)',
  filter_params: ['alice'],
).each do |t|
  puts "#{t.id} at #{t.timestamp}"
end
# endsnippet

# snippet list-account-balances
balances = chain.balances.query(
  filter: 'account_alias=$1',
  filter_params: ['alice'],
).each do |b|
  puts "Alice's balance of #{b.sum_by['asset_alias']}: #{b.amount}"
end
# endsnippet

# snippet list-account-unspent-outputs
unspentOutputs = chain.unspent_outputs.query(
  filter: 'account_alias=$1 AND asset_alias=$2',
  filter_params: ['alice', 'gold'],
).each do |u|
  puts "Output ID: #{u.id}"
end
# endsnippet

File: docs/core/examples/node/accounts.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let assetKey, aliceKey, bobKey

Promise.all([
  client.mockHsm.keys.create(),
  client.mockHsm.keys.create(),
  client.mockHsm.keys.create(),
]).then(keys => {
  assetKey = keys[0].xpub
  aliceKey = keys[1].xpub
  bobKey   = keys[2].xpub

  signer.addKey(assetKey, client.mockHsm.signerConnection)
  signer.addKey(aliceKey, client.mockHsm.signerConnection)
  signer.addKey(bobKey, client.mockHsm.signerConnection)
}).then(() => Promise.all([
  client.assets.create({
    alias: 'gold',
    rootXpubs: [assetKey],
    quorum: 1,
  }),
  client.assets.create({
    alias: 'silver',
    rootXpubs: [assetKey],
    quorum: 1,
  }),
  // snippet create-account-alice
  client.accounts.create({
    alias: 'alice',
    rootXpubs: [aliceKey],
    quorum: 1,
    tags: {
      type: 'checking',
      first_name: 'Alice',
      last_name: 'Jones',
      user_id: '12345',
    }
  })
  // endsnippet
  ,
  // snippet create-account-bob
  client.accounts.create({
    alias: 'bob',
    rootXpubs: [bobKey],
    quorum: 1,
    tags: {
      type: 'savings',
      first_name: 'Bob',
      last_name: 'Smith',
      user_id: '67890',
    }
  })
  // endsnippet
])).then(() =>
  // snippet list-accounts-by-tag
  client.accounts.queryAll({
    filter: 'tags.type=$1',
    filterParams: ['savings'],
  }, (account, next) => {
    console.log('Account ID ' + account.id + ' alias ' + account.alias)
    next()
  })
  // endsnippet
).then(() =>
  client.transactions.build(builder => {
    builder.issue({ assetAlias: 'gold', amount: 100 })
    builder.issue({ assetAlias: 'silver', amount: 100 })
    builder.controlWithAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 100
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'silver',
      amount: 100
    })
  }).then(issuance => signer.sign(issuance))
    .then(signed => client.transactions.submit(signed))
).then(() => {
  // snippet build-transfer
  const spendPromise = client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 10
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 10
    })
  })
  // endsnippet

  return spendPromise.then(spendingTx => {
    // snippet sign-transfer
    const signingPromise = signer.sign(spendingTx)
    // endsnippet

    return signingPromise
  }).then(signedSpendingTx =>
    // snippet submit-transfer
    client.transactions.submit(signedSpendingTx)
    // endsnippet
  )
}).then(() => {
  // snippet create-receiver
  const bobReceiverSerializedPromise = client.accounts.createReceiver({
    accountAlias: 'bob',
  }).then(bobReceiver => {
    return JSON.stringify(bobReceiver)
  })
  // endsnippet

  return bobReceiverSerializedPromise.then(bobReceiverSerialized =>
    // snippet transfer-to-receiver
    client.transactions.build(builder => {
      builder.spendFromAccount({
        accountAlias: 'alice',
        assetAlias: 'gold',
        amount: 10
      })
      builder.controlWithReceiver({
        receiver: JSON.parse(bobReceiverSerialized),
        assetAlias: 'gold',
        amount: 10
      })
    }).then(template => {
      return signer.sign(template)
    }).then(signed => {
      return client.transactions.submit(signed)
    })
    // endsnippet
  )
}).then(() =>
  // snippet list-account-txs
  client.transactions.queryAll({
    filter: 'inputs(account_alias=$1) AND outputs(account_alias=$1)',
    filterParams: ['alice'],
  }, (transaction, next) => {
    console.log(transaction.id + ' at ' + transaction.timestamp)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-account-balances
  client.balances.queryAll({
    filter: 'account_alias=$1',
    filterParams: ['alice'],
  }, (balance, next) => {
    console.log("Alice's balance of " + balance.sumBy.assetAlias + ': ' + balance.amount)
    next()
  })
  // endsnippet
).then(() =>
  // snippet list-account-unspent-outputs
  client.unspentOutputs.queryAll({
    filter: 'account_alias=$1 AND asset_alias=$2',
    filterParams: ['alice', 'gold'],
  }, (unspent, next) => {
    console.log('Output ID: ' + unspent.id)
    next()
  })
  // endsnippet
).catch(err =>
  process.nextTick(() => { throw err })
)
